File: src/com/lushprojects/circuitjs1/client/RailElm.java
/*    
    Copyright (C) Paul Falstad and Iain Sharp
    
    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.lushprojects.circuitjs1.client;

class RailElm extends VoltageElm {
	public RailElm(int xx, int yy) { 
		super(xx, yy, WF_DC); 

	}
	RailElm(int xx, int yy, int wf) {
		super(xx, yy, wf); 
	}

	public RailElm(int xa, int ya, int xb, int yb, int f,
			StringTokenizer st) {
		super(xa, ya, xb, yb, f, st);
	}

    
    final int FLAG_CLOCK = 1;
    int getDumpType() { return 'R'; }
    int getPostCount() { return 1; }

    void setPoints() {
	super.setPoints();
	lead1 = interpPoint(point1, point2, 1-circleSize/dn);
    }
    
    String getRailText() {
	return null;
    }
    
    void draw(Graphics g) {
	String rt = getRailText();
        double w = rt == null ? circleSize : g.context.measureText(rt).getWidth()/2;
        if (w > dn*.8)
            w = dn*.8;
	lead1 = interpPoint(point1, point2, 1-w/dn);
	setBbox(point1, point2, circleSize);
	setVoltageColor(g, volts[0]);
	drawThickLine(g, point1, lead1);
	drawRail(g);
	drawPosts(g);
	curcount = updateDotCount(-current, curcount);
	if (!isCreating())
	    drawDots(g, point1, lead1, curcount);
    }

    void drawRail(Graphics g) {
	if (waveform == WF_SQUARE && (flags & FLAG_CLOCK) != 0)
	    drawRailText(g, "CLK");
	else if (waveform == WF_DC || waveform == WF_VAR) {
	    g.setColor(needsHighlight() ? selectColor : whiteColor);
	    setPowerColor(g, false);
	    double v = getVoltage();
	    String s;
	    if (Math.abs(v) < 1)
	    	s = showFormat.format(v)+" V";
	    else
	    	s = getShortUnitText(v, "V");
	    if (getVoltage() > 0)
		s = "+" + s;
	    drawLabeledNode(g, s, point1, lead1);
	} else {
	    drawWaveform(g, point2);
	}
    }
    
    void drawRailText(Graphics g, String s) {
	g.setColor(needsHighlight() ? selectColor : whiteColor);
	setPowerColor(g, false);
	drawLabeledNode(g, s, point1, lead1);
    }
    
    double getVoltageDiff() { return volts[0]; }
    void setVoltageSource(int n, VoltageSource v) {
	super.setVoltageSource(n, v);
	if (internalResistance > 0)
	    v.setNodes(CircuitNode.ground, nodes[1]);
	else
	    v.setNodes(CircuitNode.ground, nodes[0]);
    }
    void stamp() {
	CircuitNode vsNode = internalResistance > 0 ? nodes[1] : nodes[0];
	if (waveform == WF_DC)
	    sim.stampVoltageSource(CircuitNode.ground, vsNode, voltSource, getVoltage());
	else
	    sim.stampVoltageSource(CircuitNode.ground, vsNode, voltSource);
	if (internalResistance > 0)
	    sim.stampResistor(nodes[1], nodes[0], internalResistance);
    }
    void doStep() {
	CircuitNode vsNode = internalResistance > 0 ? nodes[1] : nodes[0];
	if (waveform != WF_DC)
	    sim.updateVoltageSource(CircuitNode.ground, vsNode, voltSource, getVoltage());
    }
    boolean hasGroundConnection(int n1) { return true; }
    void addRoutingObstacle(WireRouter router) {
	router.addWire(point1.x, point1.y, point2.x, point2.y);
	router.addObstacle(point2.x - circleSize, point2.y - circleSize,
			   point2.x + circleSize, point2.y + circleSize);
    }

    int getShortcut() { return 'V'; }
    boolean validateRailNode(int n) {
	FindPathInfo fpi = new FindPathInfo(FindPathInfo.VOLTAGE, this, getNode(n), sim);
	if (fpi.findPath(CircuitNode.ground)) {
	    //sim.stop("Path to ground with no resistance!", this);
	    internalResistance = .001;
	    return false;
	}
	return true;
    }
    boolean validate() { return internalResistance > 0 || validateRailNode(0); }

    boolean getDragVertical(boolean requestedVertical) {
        return requestedVertical;
    }

//    void drawHandles(Graphics g, Color c) {
//    	g.setColor(c);
//		g.fillRect(x-3, y-3, 7, 7);
//    }

}

File: RailElm.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { VoltageElm } from "./VoltageElm";
import { CircuitElm } from "./CircuitElm";
import { CircuitNode } from "./CircuitNode";
import { FindPathInfo } from "./FindPathInfo";
import { Graphics } from "./Graphics";
import { StringTokenizer } from "./StringTokenizer";
import { WireRouter } from "./WireRouter";

export class RailElm extends VoltageElm {
    constructor(xx: number, yy: number);
    constructor(xx: number, yy: number, wf: number);
    constructor(xa: number, ya: number, xb: number, yb: number, f: number, st: StringTokenizer);
    constructor(xa: number, ya: number, xbOrWf?: number, yb?: number, f?: number, st?: StringTokenizer) {
        if (yb === undefined) {
            super(xa, ya, xbOrWf ?? VoltageElm.WF_DC);
        } else {
            super(xa, ya, xbOrWf!, yb, f!, st!);
        }
    }

    isRailElm(): boolean { return true; }
    readonly FLAG_CLOCK = 1;
    getDumpType(): number { return 'R'.charCodeAt(0); }
    getPostCount(): number { return 1; }

    setPoints(): void {
        super.setPoints();
        this.lead1 = this.interpPoint(this.point1, this.point2, 1-this.circleSize/this.dn);
    }

    getRailText(): string | null {
        return null;
    }

    draw(g: Graphics): void {
        const rt = this.getRailText();
        let w = rt == null ? this.circleSize : g.context.measureText(rt).width/2;
        if (w > this.dn*.8)
            w = this.dn*.8;
        this.lead1 = this.interpPoint(this.point1, this.point2, 1-w/this.dn);
        this.setBbox(this.point1, this.point2, this.circleSize);
        this.setVoltageColor(g, this.nodes[0].v);
        CircuitElm.drawThickLine(g, this.point1, this.lead1!);
        this.drawRail(g);
        this.drawPosts(g);
        this.curcount = this.updateDotCountImpl(-this.current, this.curcount);
        if (!this.isCreating())
            this.drawDots(g, this.point1, this.lead1!, this.curcount);
    }

    drawRail(g: Graphics): void {
        if (this.waveform == VoltageElm.WF_SQUARE && (this.flags & this.FLAG_CLOCK) != 0)
            this.drawRailText(g, "CLK");
        else if (this.waveform == VoltageElm.WF_DC || this.waveform == VoltageElm.WF_VAR) {
            g.setColor(this.needsHighlight() ? RailElm.selectColor : RailElm.whiteColor);
            this.setPowerColor(g, false);
            const v = this.getVoltage();
            let s: string;
            if (Math.abs(v) < 1)
                s = CircuitElm.showFormat.format(v)+" V";
            else
                s = CircuitElm.getShortUnitText(v, "V");
            if (this.getVoltage() > 0)
                s = "+" + s;
            this.drawLabeledNode(g, s, this.point1, this.lead1!);
        } else {
            this.drawWaveform(g, this.point2);
        }
    }

    drawRailText(g: Graphics, s: string): void {
        g.setColor(this.needsHighlight() ? RailElm.selectColor : RailElm.whiteColor);
        this.setPowerColor(g, false);
        this.drawLabeledNode(g, s, this.point1, this.lead1!);
    }

    getVoltageDiff(): number { return this.nodes[0].v; }
    setVoltageSource(n: number, v: any): void {
        super.setVoltageSource(n, v);
        if (this.internalResistance > 0)
            v.setNodes(CircuitNode.ground, this.nodes[1]);
        else
            v.setNodes(CircuitNode.ground, this.nodes[0]);
    }
    stamp(): void {
        const vsNode = this.internalResistance > 0 ? this.nodes[1] : this.nodes[0];
        if (this.waveform == VoltageElm.WF_DC)
            CircuitElm.sim.stampVoltageSource(CircuitNode.ground, vsNode, this.voltSource, this.getVoltage());
        else
            CircuitElm.sim.stampVoltageSource(CircuitNode.ground, vsNode, this.voltSource);
        if (this.internalResistance > 0)
            CircuitElm.sim.stampResistor(this.nodes[1], this.nodes[0], this.internalResistance);
    }
    doStep(): void {
        const vsNode = this.internalResistance > 0 ? this.nodes[1] : this.nodes[0];
        if (this.waveform != VoltageElm.WF_DC)
            CircuitElm.sim.updateVoltageSource(CircuitNode.ground, vsNode, this.voltSource, this.getVoltage());
    }
    hasGroundConnection(n1: number): boolean { return true; }
    addRoutingObstacle(router: WireRouter): void {
        router.addWire(this.point1.x, this.point1.y, this.point2.x, this.point2.y);
        router.addObstacle(this.point2.x - this.circleSize, this.point2.y - this.circleSize,
                           this.point2.x + this.circleSize, this.point2.y + this.circleSize);
    }

    getShortcut(): number { return 'V'.charCodeAt(0); }
    validateRailNode(n: number): boolean {
        const fpi = new FindPathInfo(FindPathInfo.VOLTAGE, this, this.getNode(n), CircuitElm.sim);
        if (fpi.findPath(CircuitNode.ground)) {
            //CircuitElm.sim.stop("Path to ground with no resistance!", this);
            this.internalResistance = .001;
            return false;
        }
        return true;
    }
    validate(): boolean { return this.internalResistance > 0 || this.validateRailNode(0); }

//    void drawHandles(Graphics g, Color c) {
//    	g.setColor(c);
//		g.fillRect(x-3, y-3, 7, 7);
//    }
}
